### ISSUE TITLE
MMX Nvidia OpenCL Support for Timelord processing

### ISSUE THREAD

**Issue by tjb-altf4** [guydavis/machinaris#533]
Hi Guy,
These are the steps I used to add OpenCL/GPU support for MMX in my own (non-machinaris) container, this will allow hardware acceleration of timelord processing.

- Nvidia docker framework (for Unraid this is the Nvidia Driver plugin)
- set nvidia env flags in dockerfile or passed in run command
```
NVIDIA_VISIBLE_DEVICES=all            <- this might need to be tuned per user system to specify device
NVIDIA_DRIVER_CAPABILITIES=compute,utility
```
- specify container is to use nvidia docker framework
`--runtime=nvidia `
- RUN or entrypoint
```
mkdir -p /etc/OpenCL/vendors && \
    echo "libnvidia-opencl.so.1" > /etc/OpenCL/vendors/nvidia.icd
```

**Comment by tjb-altf4** [comment]
you can confirm it is working by running `clinfo` in the container.
if there are multiple gpus passed, it looks like you can choose which device to use in Node.json opencl_device parameter

**Comment by tjb-altf4** [comment]
my current entrypoint script has this added:
`if [[ ${OPENCL_GPU} == 'nvidia' ]]; then    
    mkdir -p /etc/OpenCL/vendors
    echo "libnvidia-opencl.so.1" > /etc/OpenCL/vendors/nvidia.icd
fi`

so I just need to add a few ENVs to turn on support
`-e OPENCL_GPU=nvidia \
-e NVIDIA_VISIBLE_DEVICES=0 \
-e NVIDIA_DRIVER_CAPABILITIES=compute,utility \
--runtime=nvidia \`

this should give you some flexibility on supporting AMD GPU or Intel iGPU down the road

**Comment by guydavis** [comment]
@tjb-altf4 Thanks for the tips.  Would you be able to point me to the example Dockerfile build and entrypoint you are mentioning above?  For example, I'm not sure if:

1. A new base image is required or will stock Ubuntu Hirsute work?
2. If stock Ubuntu will work, what installs of apps/tools/libs/frameworks are you performing?
3. Assuming 1 & 2 above are sorted out, are just those few lines you posted above all that is truly needed in entrypoint.sh?
4. Are you building MMX into your image with particular switches to enable this?  
5. Here is the [current Machinaris build for MMX](https://github.com/guydavis/machinaris/blob/develop/scripts/forks/mmx_install.sh).  What would need to change to support those both with and without the OpenCL framework?

Thanks in advance for your expertise on this!

**Comment by tjb-altf4** [comment]
1. I'm using stock latest Ubuntu (currently focal), just needs the libraries mentioned in MMX, which your build script has by the look of it.  Nvidia's docker library is required on the host, but on the container side requirements are minimal, anything needed seems to be there by already. Hirsute should work, but I can validate against this version later on if you need to troubleshoot.
2. Pretty much nothing, just needs that OpenCL vendor file to be added either at build time or run time i.e. entrypoint.  Bare minimum is this script below, I added this to my dockerfile during my testing, and later moved it to the entrypoint script  when I knew everything was working.
```
mkdir -p /etc/OpenCL/vendors
echo "libnvidia-opencl.so.1" > /etc/OpenCL/vendors/nvidia.icd
```
3. Yep thats all mine has to enable OpenCL for Nvidia.
4. These would provide nice defaults that are harmless when not in use.
```
-e NVIDIA_VISIBLE_DEVICES=all \
-e NVIDIA_DRIVER_CAPABILITIES=compute,utility \
```
but ideally you would target the device by position or GUID, which you can later override in unraid-template or compose ENVs, e.g. `-e NVIDIA_VISIBLE_DEVICES=0` or  `-e NVIDIA_VISIBLE_DEVICES=GPU-a8b96998-xxxx-xxxx`
Further explained here: https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/user-guide.html#gpu-enumeration.
If the above defaults are in place, you only need the switch added to the container run command (extra parameters in Unraid) `--runtime=nvidia` 
5. Although you can add the OpenCL info at build stage, probably simpler to do it in an entrypoint script.  If you do a check for an ENV like the one I added called OPENCL_GPU (naming isn't significant), then your entrypoint can read these add/remove files based on how you or the user have configured the docker run.

As an example, a highly simplified version of my docker run with OpenCL looks like this (remembering my example entrypoint has a check for OPENCL_GPU)
```
docker run --rm \
  --name mmx-node-blockchain \
  -e OPENCL_GPU=nvidia \
  --runtime=nvidia \
  mmx-node-blockchain:testnet4
```
and no OpenCl looks like this
```
docker run --rm \
  --name mmx-node-blockchain \
  mmx-node-blockchain:testnet4
```

**Comment by tjb-altf4** [comment]
Hey @guydavis,
Updates look good, just a few notes.
In FAQ, you have the following:
> For Unraid, create 3 new Variables (shown) and add runtime=nvidia to the "Extra Parameters" field.

this should be (leading dashes are needed):
> For Unraid, create 3 new Variables (shown) and add --runtime=nvidia to the "Extra Parameters" field.

I'd also make mention that the host needs Nvidia's docker framework installed.
For unraid users this as simple as installing the Nvidia Driver plugin.

For other OS I've not set it up before, but I think Nvidia's documentation should help those users, WSL might be more difficult though.
https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/user-guide.html#adding-the-nvidia-runtime

**Comment by guydavis** [comment]
Available in version 0.6.9.  Thanks for the all the help on this enhancement!
